test/tests/layer.js

```diff
@@ -82,7 +89,30 @@ suite("d3.layer", function() {
 			this.layer.draw([]);
 			assert(this.insert.calledOn(this.dataBind.returnValues[0].enter.returnValues[0]));
 		});
-		
+		test("by default removes exiting nodes from the DOM", function() {
+			this.layer.draw([1]);
+			assert.equal(this.layer.selectAll("g").size(), 1);
+			this.layer.draw([]);
+			assert.equal(this.layer.selectAll("g").size(), 0);
+		});
+		test("invokes the provided `remove` method in the context of the layer's bound 'exit' selection", function() {
+			var updating, exiting;
+
+			this.layerWithRemove.draw([1, 2, 3]);
+			this.layerWithRemove.draw([]);
+
+			updating = this.dataBind.returnValues[1];
+			exiting = updating.exit.returnValues[0];
+
+			assert(this.remove.calledOn(exiting));
```

Comment:
In the issue that prompted this change (https://github.com/misoproject/d3.chart/issues/63), we identified an asymmetry between how data-bound elements are inserted and how they are removed. In the former case, d3.chart requires a function dedicated to the task of insertion. In the latter case, users are expected to remove data-bound elements in a lifecycle event.  `Layer#remove` is an attempt to formalize the way charts define "cleanup" logic for data-bound elements. It does not remove the layer wholesale, though--this is distinct from `d3.selection.prototype.remove`.

In light of that, I don't believe this patch suffers from the problem you are describing here. Can you confirm? If so, I think this is indicative of a confusing method name, and we should talk about finding a better name!